src/Symfony/Bundle/FrameworkBundle/DataCollector/MessagesDataCollector.php

```diff
@@ -0,0 +1,92 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\FrameworkBundle\DataCollector;
+
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\DataCollector\DataCollector;
+use Symfony\Component\Message\MiddlewareInterface;
+
+/**
+ * @author Samuel Roze <samuel.roze@gmail.com>
+ */
+class MessagesDataCollector extends DataCollector implements MiddlewareInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function collect(Request $request, Response $response, \Exception $exception = null)
+    {
+        return $this->data;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getName()
+    {
+        return 'messages';
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function reset()
+    {
+        $this->data = array();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function handle($message, callable $next)
+    {
+        $debugRepresentation = array(
+            'message' => array(
+                'type' => get_class($message),
+            ),
+        );
+
+        try {
+            $result = $next($message);
+
+            if (is_object($result)) {
+                $debugRepresentation['result'] = array(
+                    'type' => get_class($result),
+                );
+            } else {
+                $debugRepresentation['result'] = array(
+                    'type' => gettype($result),
+                    'value' => $result,
+                );
+            }
+        } catch (\Throwable $exception) {
```

Comment:
Do we really want to catch any `\Error`?

Reply:
I'd say so. If your handler fails because of a PHP typo, you'll be super happy that your receiver re-queue your message :)

Reply:
Re-queueing would mean creating a loop with the consumers. Do we have sensible defaults when setting up a queue to avoid a redelivering loop that would fill up the consumers?

Reply:
In this specific context, sending the exception back to the `Generator` is to ensure that such behaviour (requeuing) could be done by the adapters. 

The way to handle the case you are talking about (i.e. an exception looping the messages over and over) really depends on the way the messages are managed. In [this "native" AMQP adapter](https://github.com/sroze/symfony/blob/9b9907b2cdb66625921a3b1844f9b88810e2b453/src/Symfony/Bridge/PhpAmqp/AmqpReceiver.php#L51-L52), you can see that an interface is exposed for you to implement it on exceptions that would mean "reject this message" (i.e. don't requeue it). 

This approach (explicitly non-requeuing some message) is the best to prevent any lose on Symfony's users architectures. You don't want a message to be lost if something wrong happened because a 3rd party API was down for 2 minutes for example.